dali/pipeline/operators/reader/coco_reader_op.cc

```diff
@@ -14,47 +14,446 @@
+  }
+  return detail::load_filenames(meta_files_path + "/filenames.txt");
+}
+
+void COCOReader::ValidateOptions(const OpSpec &spec) {
+  DALI_ENFORCE(
+    spec.HasArgument("meta_files_path") || spec.HasArgument("annotations_file"),
+    "`meta_files_path` or `annotations_file` must be provided");
+  DALI_ENFORCE(
+    !spec.HasArgument("file_list"),
+    "Argument `file_list` is no longer supported for `COCOReader`."
+    "The same functionality can be implemented with meta files option.");
+  DALI_ENFORCE(!skip_cached_images_,
+    "COCOReader doesn't support `skip_cached_images` option");
+
+  if (spec.HasArgument("meta_files_path")) {
+    DALI_ENFORCE(
+      !spec.HasArgument("skip_empty"),
+      "When reading data from meta files `skip_empty` is not working.");
+    DALI_ENFORCE(
+      !spec.HasArgument("ratio"),
+      "When reading data from meta files `ratio` is not working.");
+    DALI_ENFORCE(
+      !spec.HasArgument("ltrb"),
+      "When reading data from meta files `ltrb` is not working.");
+    DALI_ENFORCE(
+      !spec.HasArgument("size_threshold"),
+      "When reading data from meta files `size_threshold` is not working.");
+    DALI_ENFORCE(
+      !spec.HasArgument("dump_meta_files"),
+      "When reading data from meta files `dump_meta_files` is not working.");
+    DALI_ENFORCE(
+      !spec.HasArgument("dump_meta_files_path"),
+      "When reading data from meta files `dump_meta_files_path` is not working.");
+  }
+
+  if (spec.HasArgument("dump_meta_files")) {
+    DALI_ENFORCE(
+      spec.HasArgument("dump_meta_files_path"),
+      "When dumping meta files `dump_meta_files_path` must be provided.");
+  }
+}
+
+std::vector<std::pair<std::string, int>> COCOReader::ParseJsonAnnotations(const OpSpec &spec) {
+  std::vector<detail::ImageInfo> image_infos;
+  std::vector<detail::Annotation> annotations;
+  std::map<int, int> category_ids;
+
+  detail::parse_json_file(
+    spec,
+    image_infos,
+    annotations,
+    category_ids);
+
+  bool skip_empty = spec.GetArgument<bool>("skip_empty");
+  bool ratio = spec.GetArgument<bool>("ratio");
+
+  std::sort(image_infos.begin(), image_infos.end(), [](auto &left, auto &right) {
+    return left.original_id_ < right.original_id_;});
+  std::stable_sort(annotations.begin(), annotations.end(), [](auto &left, auto &right) {
+    return left.image_id_ < right.image_id_;});
+
+  detail::Annotation sentinel;
+  sentinel.image_id_ = -1;
+  annotations.emplace_back(std::move(sentinel));
+
+  int new_image_id = 0;
+  int annotation_id = 0;
+  int total_count = 0;
+
+  ImageIdPairs image_id_pairs;
+
+  for (auto &image_info : image_infos) {
+    int objects_in_sample = 0;
+    while (annotations[annotation_id].image_id_ == image_info.original_id_) {
+      const auto &annotation = annotations[annotation_id];
+      labels_.emplace_back(category_ids[annotation.category_id_]);
+      if (ratio) {
+        boxes_.emplace_back(annotation.box_[0] / static_cast<float>(image_info.width_));
```

Comment:
No need for all these cast - `box_` is already an array of floats.

Reply:
Also: why not `push_back` for a simple type?

Reply:
Done